server/src/main/java/org/elasticsearch/index/seqno/RetentionLeaseSyncAction.java

```diff
@@ -88,6 +94,46 @@ public RetentionLeaseSyncAction(
                 ThreadPool.Names.MANAGEMENT, false);
     }
 
+    @Override
+    protected void doExecute(Task parentTask, Request request, ActionListener<Response> listener) {
+        // Skip reroute phase as we are on the primary shard already.
+        final IndexShard indexShard = getIndexShard(request.shardId());
+        final ShardRouting shardRouting = indexShard.routingEntry();
+        if (shardRouting.primary() == false) {
+            throw new ShardNotInPrimaryModeException(indexShard.shardId(), indexShard.state());
```

Comment:
Apart from the change in this exception, these pre-flight checks seem to duplicate the ones performed in `AsyncPrimaryAction#doRun` which are called anyway as part of the handling of the transport request we send below. Furthermore I think the difference in the exception is not watertight - it's still possible for us to get a `RetryOnPrimaryException` isn't it?

Could we instead just send the child request and adjust any `RetryOnPrimaryException` received in response?

I'm also not convinced that `doExecute` is the right place to do this. Could we have a separate method that explicitly bypasses the reroute phase like in `TransportResyncReplicationAction`? We could reasonably assert that `doExecute` isn't called.

Reply:
++. I pushed https://github.com/elastic/elasticsearch/pull/48430/commits/6e4500f8aff11772d50155fb351047e20911b6e3.